### PR TITLE
Updated Function Browser and related files

### PR DESCRIPTION
**Pull request by Matthew-Andrew**
**Description of work.**

This is the changes from #26524 related to the core fitting framework and not the IndirectFitting GUI

**To test:**
* Code review
* Check that the existing usages of the FunctionBrowser still work i.e. in plot based fitting, ALC interface, IndirectAnalysis and Muon Analysis

Fixes (partially) #22055  

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by alicerussell1 — CHANGES_REQUESTED** [review]
This looks good, the interfaces that use the function browser all seem to still work the same as before. Just a couple of comments about the code

File: qt/widgets/common/src/FunctionBrowser.cpp [hunk 1]
```diff
@@ -252,11 +254,43 @@ void FunctionBrowser::setLocalParameterTie(const QString &parName, int i,
...
+    auto errorColumn = paramTable.getColumn((name + "_Err").toStdString());
+    if (nRows > 1) {
+      for (size_t i = 0; i < nRows; ++i) {
+        setLocalParameterValue(name, static_cast<int>(i),
+                               valueColumn->toDouble(i),
+                               errorColumn->toDouble(i));
+      }
+    } else {
+      auto const i = getCurrentDataset();
+      setLocalParameterValue(name, static_cast<int>(i),
```
[the remark this answers is not in the document]
Comment (Mantid-Matthew) [reply qt/widgets/common/src/FunctionBrowser.cpp]: I don't think there is no.

File: qt/widgets/common/test/FunctionMultiDomainPresenterTest.h [hunk 2]
```diff
@@ -62,6 +62,8 @@ class MockFunctionView : public IFunctionView {
     }
   }
 
+  void setParameterConstraint(const QString &, const QString &) override {}
```
[the remark this answers is not in the document]
Comment (Mantid-Matthew) [reply qt/widgets/common/test/FunctionMultiDomainPresenterTest.h]: As this function does nothing in the mock adding names to the parameters leads to unused variable compiler warnings.

File: qt/widgets/common/src/FunctionBrowser/FunctionBrowserUtils.cpp [hunk 3]
```diff
@@ -56,5 +60,66 @@ std::pair<QString, int> splitFunctionPrefix(const QString &prefix) {
...
+  QString paramName;
+  Mantid::API::Expression expr;
+  expr.parse(constraint.toStdString());
+  if (expr.name() != "==") {
+    return error;
+  }
+  if (expr.size() == 3) { // lower < param < upper
+    try {
+      // check that the first and third terms are numbers
+      double d1 = boost::lexical_cast<double>(expr[0].str());
```
[the remark this answers is not in the document]
Comment (Mantid-Matthew) [reply qt/widgets/common/src/FunctionBrowser/FunctionBrowserUtils.cpp]: I don't think so no. I've added some tests and this seems to fulfill it's purpose with the assignment removed so have removed it.

File: qt/widgets/common/test/FunctionMultiDomainPresenterTest.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (alicerussell1) [reply qt/widgets/common/test/FunctionMultiDomainPresenterTest.h]: Ah I hadn't realized it was part of the mock, would it give warnings if you put commented out names like `/*name*/`?

**Review by alicerussell1 — APPROVED** [review]
This looks good to me 👍

**Pull request merged**